registry/api/errcode/errors.go

```diff
@@ -0,0 +1,206 @@
+package errcode
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+// ErrorCode represents the error type. The errors are serialized via strings
+// and the integer format may change and should *never* be exported.
+type ErrorCode int
+
+// ErrorDescriptor provides relevant information about a given error code.
+type ErrorDescriptor struct {
+	// Code is the error code that this descriptor describes.
+	Code ErrorCode
+
+	// Value provides a unique, string key, often captilized with
+	// underscores, to identify the error code. This value is used as the
+	// keyed value when serializing api errors.
+	Value string
+
+	// Message is a short, human readable decription of the error condition
+	// included in API responses.
+	Message string
+
+	// Description provides a complete account of the errors purpose, suitable
+	// for use in documentation.
+	Description string
+
+	// HTTPStatusCode provides the http status code that is associated with
+	// this error condition.
+	HTTPStatusCode int
+}
+
+var (
+	errorCodeToDescriptors = map[ErrorCode]ErrorDescriptor{}
+	idToDescriptors        = map[string]ErrorDescriptor{}
+)
+
+const (
+	// ErrorCodeUnknown is a catch-all for errors not defined below.
+	ErrorCodeUnknown ErrorCode = 10000 + iota
+)
+
+var errorDescriptors = []ErrorDescriptor{
+	{
+		Code:    ErrorCodeUnknown,
+		Value:   "UNKNOWN",
+		Message: "unknown error",
+		Description: `Generic error returned when the error does not have an
+									        API classification.`,
+		HTTPStatusCode: http.StatusInternalServerError,
+	},
+}
+
+// LoadErrors will register a new set of Errors into the system
+func LoadErrors(errs *[]ErrorDescriptor) {
```

Comment:
This should be done with each error at a time, rather than using slices of descriptors:

```
func Register(group string, desc ErrorDescriptor) ErrorCode
```

Then, when declaring the error in a package, one can have this:

```
var ErrMyErrorCode = errcode.Register("some package", errcode.Descriptor{
    Value: "MY_ERROR_CODE",
    Message: "my error happened",
    Description: "this error happens under my conditions",
    HTTPStatusCode: 400, // only fill this in if error has associated status code.
})
```

The above can fill in the error code and return it. Internally, it can check uniqueness of the "value", fill in the code of the descriptor and register it to the group.

Two queries should be supported, at the package level:

```
// GetErrorCode returns the value by the string error code. `ErrUnknown` will be returned if the error is not known.
func GetErrorCode(value string) ErrorCode

// GetErrorCodeGroup returns the named group of error descriptors
func GetErrorCodeGroup(name string) []ErrorDescriptor
```

Other queries may be necessary.

This approach has a number of benefits. First, it allows us to maintain a single registry of errors by declaration group ("distribution", "networking", etc.). It also ensures that the declarations are done near the code that uses it, ensuring that documentation and messages are appropriately updated. We also ensure uniqueness of the API code used throughout the system. We can generate error lists for documentation of the API automatically.